packages/flutter/test/material/app_bar_sliver_test.dart

```diff
@@ -2274,6 +2322,53 @@ void main() {
     expect(navToolBar.middleSpacing, NavigationToolbar.kMiddleSpacing);
   });
 
+  testWidgets('SliverAppBar should update child before child build', (WidgetTester tester) async {
+    // Regression test for https://github.com/flutter/flutter/issues/158158.
+    final List<Tab> tabs = <Tab>[];
+
+    await tester.pumpWidget(
+      MaterialApp(
+        home: StatefulBuilder(
+          builder: (BuildContext context, StateSetter setState) {
+            return DefaultTabController(
+              length: tabs.length,
+              child: Scaffold(
+                body: CustomScrollView(
+                  slivers: <Widget>[
+                    SliverAppBar(
+                      actions: <Widget>[
+                        TextButton(
+                          child: const Text('Add Tab'),
+                          onPressed: () {
+                            setState(() {
+                              tabs.add(Tab(text: 'Tab ${tabs.length + 1}'));
+                            });
+                          },
+                        ),
+                      ],
+                      bottom: _TestLifecycleTabBar(tabs: tabs),
```

Comment:
One of our [style guidelines](https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#make-each-test-entirely-self-contained) is "make each test entirely self-contained".

It's primarily for preventing accidental shared state between tests, but there are a few other reasons why we might want to factor out the `_TestLifecycleTabBar` class:

- If we can avoid declaring a private class at the top of the file that's only used for one test, it will make this library easier to understand & navigate.
- Tests are easier to debug when the failure is found in an `expect()` call, rather than an `assert` in a widget's build method.
- It'd be nice to test the `TabBar` widget, since that's what will most likely be used in the real world.

<br>

<details> <summary>Here's an idea for how we might restructure this test!</summary>

```dart
  // Regression test for https://github.com/flutter/flutter/issues/158158.
  testWidgets('SliverAppBar should update child before child build', (WidgetTester tester) async {
    List<Tab> tabs = <Tab>[];
    late StateSetter setState;
    late TabController controller;
    bool customTabBar = false;

    PreferredSizeWidget tabBar() {
      if (customTabBar) {
        return PreferredSize(
          preferredSize: const Size.fromHeight(kToolbarHeight),
          child: Builder(
            builder: (BuildContext context) {
              controller = DefaultTabController.of(context);
              return Row(children: tabs);
            },
          ),
        );
      }
      return TabBar(tabs: tabs);
    }

    await tester.pumpWidget(
      MaterialApp(
        home: StatefulBuilder(
          builder: (BuildContext context, StateSetter stateSetter) {
            setState = stateSetter;

            // Reset this widget's state when switching to a different TabBar.
            final Key key = ValueKey<bool>(customTabBar);
            return DefaultTabController(
              key: key,
              length: tabs.length,
              child: Scaffold(
                body: CustomScrollView(
                  slivers: <Widget>[
                    SliverAppBar(
                      actions: <Widget>[
                        TextButton(
                          child: const Text('Add Tab'),
                          onPressed: () {
                            setState(() {
                              tabs.add(Tab(text: 'Tab ${tabs.length + 1}'));
                            });
                          },
                        ),
                      ],
                      bottom: tabBar(),
                    ),
                  ],
                ),
              ),
            );
          },
        ),
      ),
    );

    // Initializes with zero tabs.
    expect(find.text('Tab 1'), findsNothing);
    expect(find.text('Tab 2'), findsNothing);

    await tester.tap(find.text('Add Tab'));
    await tester.pumpAndSettle();

    // No crash after tab is added.
    expect(find.text('Tab 1'), findsOneWidget);
    expect(find.text('Tab 2'), findsNothing);

    setState(() {
      tabs = <Tab>[];
      customTabBar = true;
    });

    await tester.pumpAndSettle();

    // Initializes with zero tabs.
    expect(tester.takeException(), isNull);
    expect(find.text('Tab 1'), findsNothing);
    expect(find.text('Tab 2'), findsNothing);
    expect(controller.length, tabs.length);

    await tester.tap(find.text('Add Tab'));
    await tester.pumpAndSettle();
    
    // No crash after tab is added.
    expect(tester.takeException(), isNull);
    expect(find.text('Tab 1'), findsOneWidget);
    expect(find.text('Tab 2'), findsNothing);
    expect(controller.length, tabs.length);
  });
```

</details>

Reply:
@nate-thegrate Thanks for your suggestions for the test. I tried it, but I don't know how to improve it continue.

1. Maybe we should take the exception check here.

```diff
 // No crash after tab is added.
 expect(find.text('Tab 1'), findsOneWidget);
 expect(find.text('Tab 2'), findsNothing);
+expect(tester.takeException(), isNull);

 setState(() {
   tabs = <Tab>[];
   customTabBar = true;
 });
```

2. The custom `TabBar` is declared top-level because the core reason for the issue is that `TabBar`'s lifecycle should be changed.  Actually, includes a called `DefaultTabController.of(context)` there are other necessary steps to reproduce it. We should update the tab's count in `didUpdateWidget` otherwise it's always equal.

```dart
expect(tester.takeException(), isNull);
expect(find.text('Tab 1'), findsOneWidget);
expect(find.text('Tab 2'), findsNothing);
expect(controller.length, tabs.length); // Always passed even has not current fix.
```

3. By the way. Should I offer more information on the issue somewhere? I was written some short descriptions in this PR on the top. Maybe like following are better.

Before:

```dart
flutter: _TestLifecycleTabBar.build: _tabs = 0, widget.tabs.length = 0
flutter: _TestLifecycleTabBar.build: _tabs = 0, widget.tabs.length = 0
flutter: _TestLifecycleTabBar.didUpdateWidget: _tabs = 0, widget.tabs.length = 0
flutter: _TestLifecycleTabBar.build: _tabs = 0, widget.tabs.length = 0
flutter: _TestLifecycleTabBar.build: _tabs = 0, widget.tabs.length = 1 // Crashed here.
flutter: _TestLifecycleTabBar.didUpdateWidget: _tabs = 1, widget.tabs.length = 1
flutter: _TestLifecycleTabBar.build: _tabs = 1, widget.tabs.length = 1
```

After:

```dart
flutter: _TestLifecycleTabBar.build: _tabs = 0, widget.tabs.length = 0
flutter: _TestLifecycleTabBar.didUpdateWidget: _tabs = 0, widget.tabs.length = 0
flutter: _TestLifecycleTabBar.build: _tabs = 0, widget.tabs.length = 0
flutter: _TestLifecycleTabBar.didUpdateWidget: _tabs = 1, widget.tabs.length = 1 // Update `_tabs` before `build`.
flutter: _TestLifecycleTabBar.build: _tabs = 1, widget.tabs.length = 1
```

Reply:
Thanks for following up on this—and sorry for the delay, it's been a busy week!

In my personal opinion: it's probably not necessary to compare `controller.length` with `tabs.length`.
But I'm open to hearing other perspectives :)

<br>

> 1. Maybe we should take the exception check here.
> 
> ```diff
>  // No crash after tab is added.
>  expect(find.text('Tab 1'), findsOneWidget);
>  expect(find.text('Tab 2'), findsNothing);
> +expect(tester.takeException(), isNull);
> 
>  setState(() {
>    tabs = <Tab>[];
>    customTabBar = true;
>  });
> ```

Absolutely—I totally missed that, thanks!

Reply:
@nate-thegrate If unnecessary to compare controller.length with tabs.length, the custom TabBar is also unnecessary. So I removed it and just kept TabBar for the test.